src/compiler/checker.ts

```diff
@@ -18196,6 +18196,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             if (objectType.flags & (TypeFlags.Any | TypeFlags.Never)) {
                 return objectType;
             }
+            if (indexType === neverType) {

```

Comment:
I think you can try doing something like we have in https://github.dev/microsoft/TypeScript/blob/3de2cdce4d62fab6c8fb9e8a3f78eb33125d6cd5/src/compiler/checker.ts#L17574-L17575,
and do something like `if (indexType.flags & TypeFlags.Never) { return indexType === silentNeverType ? silentNeverType : neverType; }`

Reply:
@gabritto may I ask you to run the top 300 tests agaib? I'd like to see if the current version is enough or if we need to handle the `silentNeverType` as you have suggested here. I checked `type Result = {}[never]` plus some friends in the playground and I haven't found any hiccup.

Reply:
The issue has arised because the meaning of this has changed `[silentNeverType][silentNeverType]`. It's true that the suggested change would keep the old behavior and keep "propagating" the `silentNeverType` but there are two other scenarios to consider here:
- what should be the evaluation of `[T][silentNeverType]`? Is it `silentNeverType` or `T`? It's hard to reason about `silentNeverType` because it's an internal type so we'd have to deeply understand how it is used to determine which is better. I proposed the `indexType === neverType` check specifically to avoid this - erring on the safer side to avoid breaking changes related to those complex scenarios.
- there are also other special never types internally. I thought that perhaps it would be better to keep the old behavior for them too but for example this one is quite scary and I don't think it's the right behavior:
```ts
const obj = {
  arr: [],
};

const objWithIndex: Record<string, boolean> = {};

const el = obj.arr[0];
const result = objWithIndex[el];
//    ^? const result: boolean
```

I'm not sure why in the example above `implicitNeverType[]` gets created and not `[]`.